tools/job-cache/job-cache.cpp

```diff
@@ -0,0 +1,626 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Open Group Base Specifications Issue 7
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <sqlite3.h>
+
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <json/json5.h>
+
+// This header contains useful information that you might want
+// when running a deamon
+static void log_header(FILE *file) {
+  int pid = getpid();
+  time_t t = time(NULL);
+  struct tm tm = *localtime(&t);
+  fprintf(file, "[pid=%d, %d-%02d-%02d %02d:%02d:%02d] ", pid,
+          tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday, tm.tm_hour, tm.tm_min,
+          tm.tm_sec);
+}
+
+// A generic logging function
+template <class... Args> static void log_info(Args &&...args) {
+  log_header(stdout);
+  fprintf(stdout, args...);
+  fprintf(stdout, "\n");
+  fflush(stdout);
+}
+
+// A logging function for logging and then exiting with
+// a failure code.
+template <class... Args> static void log_fail(Args &&...args) {
+  log_header(stderr);
+  fprintf(stderr, args...);
+  fprintf(stderr, "\n");
+  fflush(stderr);
+  exit(1);
+}
+
+// A logging function for logging and then exiting successfully.
+template <class... Args> static void log_exit(Args &&...args) {
+  log_header(stdout);
+  fprintf(stdout, args...);
+  fprintf(stdout, "\n");
+  fflush(stdout);
+  exit(0);
+}
+
+// Helper that only returns successful file opens and exits
+// otherwise.
+static int open_fd(const char *str, int flags) {
+  int fd = open(str, flags);
+  if (fd == -1) {
+    log_fail("open(%s): %s", str, strerror(errno));
+  }
+  return fd;
+}
+
+// Helper that only returns successful file opens and exits
```

Comment:
They're actually not the same, one passes a mode to open and the other doesn't. I *think* mode is just set to zero in the other overload but I don't know for sure so I wrote it this way. It could be that in the first overload of open it sets the mode in some different way depending on which flags you use.

Reply:
ah, my bad